scripts/delete-old-app-engine-versions.sh

```diff
@@ -0,0 +1,181 @@
+    exit 1
+}
+
+# print out error to stderr and exit
+abort() {
+    printf "${RED}ABORT: %s${RST}\n" "$1" >&2
+    exit 1
+}
+
+# always reset gcloud project regardless of error or exit
+reset() {
+    gcloud config set project "${OLD_PROJECT}" 2>/dev/null
+}
+
+# ensure that date from GNU coreutils is installed
+check_gnu_date_installed() {
+    if ! date --version 1>/dev/null 2>&1; then
+        abort "date from GNU coreutils is required"
+    fi
+}
+
+# switch gcloud project to appropriate terra ui environment
+switch_gcloud_project() {
+    gcloud config set project "${NEW_PROJECT}" 2>/dev/null
+    printf "${INFO} Switched the project to ${GRN}%s${RST}\n" "${NEW_PROJECT}"
+}
+
+# ensure that user has appropriate permissions for app engine
+check_user_permissions() {
+    if ! gcloud app services list 1>/dev/null 2>&1; then
+        GCLOUD_USER=$(gcloud config get-value account)
+        abort "User ${GCLOUD_USER} does not have permissions in ${NEW_PROJECT}"
+    fi
+}
+
+# ensure that the deletion date is always older than the oldest pr date
+set_dev_deletion_date() {
+    OLDEST_PR=$(gcloud app versions list --filter="pr-" --sort-by="LAST_DEPLOYED" | tail -n +2 | head -n 1 | sed 's/ \+/ /g')
+    OLDEST_PR_NAME=$(echo "${OLDEST_PR}" | cut -d' ' -f2)
+    OLDEST_PR_DATE=$(echo "${OLDEST_PR}" | cut -d' ' -f4)
+    OLDEST_PR_DATE=$(date --date="${OLDEST_PR_DATE}" +%F)
+
+    printf "${INFO} ${GRN}%s${RST} is the oldest PR and was deployed on ${GRN}%s${RST}\n" "${OLDEST_PR_NAME}" "${OLDEST_PR_DATE}"
+
+    UNIX_EPOCH_DELETION_DATE=$(date --date="${DELETION_DATE}" +%s)
+    UNIX_EPOCH_OLDEST_PR_DATE="$(date --date="${OLDEST_PR_DATE}" +%s)"
+
+    if [ "${UNIX_EPOCH_DELETION_DATE}" -gt "${UNIX_EPOCH_OLDEST_PR_DATE}" ]; then
+        DELETION_DATE=$(date --date="${OLDEST_PR_DATE} -1 day" +%F)
+    fi
+}
+
+# return versions of app engine that match filter
+filter_app_engine_versions() {
+    gcloud app versions list --filter="$1" 2>/dev/null | tail -n +2 | sed 's/ \+/ /g' | cut -d' ' -f2
+}
+
+# ensure that deletions leave a certain number of deployments
+check_remaining_items() {
+    REMAIN_LIST_ITEMS=($(filter_app_engine_versions "version.createTime.date('%Y-%m-%d', Z)>'${DELETION_DATE}'"))
+    REMAIN_LIST_COUNT="${#REMAIN_LIST_ITEMS[@]}"
+    if [ "${REMAIN_LIST_COUNT}" -lt 1 ]; then
+        abort "all deployments would be deleted"
+    elif [ "$REMAIN_LIST_COUNT" -lt 15 ]; then
+        abort "less than 15 deployments would remain"
+    fi
+}
+
+# ensure that deletions erase a certain number of deployments
+check_deletion_items() {
+    DELETE_LIST_ITEMS=($(filter_app_engine_versions "version.createTime.date('%Y-%m-%d', Z)<='${DELETION_DATE}'"))
+    DELETE_LIST_COUNT="${#DELETE_LIST_ITEMS[@]}"
+    if [ "${DELETE_LIST_COUNT}" -lt 1 ]; then
+        abort "no deployments to delete"
+    elif [ "${DELETE_LIST_COUNT}" -lt 30 ]; then
+        abort "less than 30 deployments to delete"
```

Comment:
Why abort if less than 30 deployments would be deleted? Should those deployments not be deleted?

Reply:
I thought the goal was to automate this to run on a regular basis? In that case, it seems like this would produce spurious failures at times where there happened to only be a few deployments.